### ISSUE TITLE
Split dependency resolution into resolve and downloading (lock dependency)

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#2989]
### steps

1. Run `update` task (or `compile` task that calls `update`) on a newly created project.

### expectation

1. `update` internally calls `resolveDependencies`.
2. `resolveDependencies` performs Ivy resolution and persists the resulting dependency graph into a JSON file under `src/sbt/resolveDependencies.json`. This file might contain:
  a. SHA-1 of the input `libraryDependencies`, and SHA-1 of all `libraryDependencies` across the build ("build clock").
  b. URLs for *all artifacts*. Or an expression relative to resolvers.
  c. Timestamp for `-SNAPSHOT` resolution.
3. Using the persisted dependency graph, `update` task downloads *required artifacts*.

### steps 2

1. The user git commits `src/sbt/resolveDependencies.json`.
2. Later day, some other user Bob git clones and runs `compile` task (that internally calls `update`).

### expectation 2

4. `compile` internally calls `update`, which internally calls `resolveDependencies`.
5. `resolveDependencies` finds that there's a persisted dependency graph at `src/sbt/resolveDependencies.json`, and returns it without calling Ivy.
6.  Using the returned dependency graph, `update` task downloads *required artifacts*.

### steps 3

1. Bob adds `commons-io` 1.0 to `libraryDependencies`.
2. Bob reloads the build.
3. Bob runs `compile`.

### expectation 3

7. `compile` internally calls `update`, which internally calls `resolveDependencies`.
8. Detecting the change in `libraryDependencies`, which changes the SHA-1 of all libraryDependencies  ("build clock"), all `resolveDependencies.json` are invalidated, and performs Ivy resolution and overwrites the file.

### note

This should work similar to `Gemfile.lock`.

In addition, by using our own caching mechanism, we should be able to avoid global Ivy locking when Ivy is not invoked. Note that it will need special handling for `-SNAPSHOT`.

Ref https://github.com/sbt/sbt/issues/290

**Comment by dwijnand** [comment]
Please could you describe the issue with steps/problem/expectation?

Does it mean that the global lock file in the ivy cache it will only lock the cache when it's downloading and not when it's just resolving locally?

Or is this some project-local freeze/lock file equivalent? (e.g Gemfile.lock)

**Comment by eed3si9n** [comment]
@dwijnand Updated the description. The latter, but there's some impact to Ivy locking as well.

**Comment by stephennancekivell** [comment]
If I understand you correctly, are you suggesting sbt creates a new file `src/sbt/resolveDependencies.json` as a cache of the dependency resolution.

I like the idea, but couldnt this be put under 'target/' somewhere and hidden from the user. I dont know if devs want another thing to worry about.

If the goal is to improve startup times for one developer. Putting it in `target` hidden from source control is ideal.

Or if the goal is to share the performance gain amongst a team, then `src` is better.


Another concern which Im not sure about..
Could dependency resolution be different for different users legitimately? Some users might have different resolvers because of performance/location or security/firewall reasons. In which case saving the resolution in `src` would be bad.

**Comment by eed3si9n** [comment]
The goal is both for performance improvement and consistency across multiple users.

> Could dependency resolution be different for different users legitimately? Some users might have different resolvers because of performance/location or security/firewall reasons. In which case saving the resolution in src would be bad.

If the list of resolvers do not match up, sbt or the user can rerun `resolveDependencies`.

**Comment by stephennancekivell** [comment]
Good point, thanks.

**Comment by nafg** [comment]
Wouldn't a more natural location be under `project/`, alongside `build.properties`?

**Comment by eed3si9n** [comment]
> Wouldn't a more natural location be under project/, alongside build.properties?

@nafg I think of `project/` as a space for metabuild, or something that affects the metabuild.
If not `src/sbt`, the only logical place would be the root directory along side `build.sbt`. Also note there might be multiple of these files per subproject.

**Comment by nafg** [comment]
@eed3si9n that's refuted by `build.properties` and `*.scala` in `project/`

**Comment by eed3si9n** [comment]
How so? `project/Foo.scala` is a source code of the root project of the metabuild.
`project/build.properties` affects the library dependency of the metabuild, which is the sbt itself's version.

**Comment by nafg** [comment]
Ok I misunderstood what you mean by meta build, I thought you meant like `project/*.sbt`. I guess you would call that the meta-meta-build, and "build" would be `./*.scala` (and `src/main/scala` ...)?

Anyway if that's what you meant then I don't see the problem with putting it there?

Put another way, the [meta] build has a few input files, including `./*.sbt`, `project/*.scala`, and `project/dependencies.lock` (or whatever you want to call the file). I still don't see the problem, not on a technical level nor in terms of intuition.

**Comment by eed3si9n** [comment]
Consider a (proper) build that exists on the `~/hello/`, and is a set of subprojects `root`, `app`, and `library`. If we say the lock files go to the `project/` then what would each be called? Also suppose `~/hello/project/` contains `plugin.sbt` pointing to sbt-assembly. Where would the lock file go for the plugin resolution (i.e. library resolution for metabuild)?

**Comment by nafg** [comment]
Why not look at the whole build (all subprojects) together? For instance what if two subprojects both have a dependency, and one depends on the other -- the dependency in one can evict the dependency in the other. Or does that not fit with the way resolution works?

As for the meta build, why not do what is already done and use `project/project`?

`src/sbt` just seems very ad-hoc. Does anything else use it? Will anything else use it?

Another option is to go more in line with what other build tools do, and have a file in the main directory. Perhaps start with a dot if concerned about clutter, e.g., `.sbt-deps.lock`.

**Comment by eed3si9n** [comment]
Updated expectation and added steps 3 and expectation 3 based on discussion with @jvican:

### steps 3

1. Bob adds `commons-io` 1.0 to `libraryDependencies`.
2. Bob reloads the build.
3. Bob runs `compile`.

### expectation 3

7. `compile` internally calls `update`, which internally calls `resolveDependencies`.
8. Detecting the change in `libraryDependencies`, which changes the SHA-1 of all libraryDependencies  ("build clock"), all `resolveDependencies.json` are invalidated, and performs Ivy resolution and overwrites the file.

**Comment by jvican** [comment]
Can someone assign me to this ticket?

**Comment by ktham** [comment]
In which milestone is this PR shooting for?

**Comment by ritschwumm** [comment]
while i do see the point of having a reproducable build - please, please, please do _not_ put generated files into `src/`. this folder should be reserved for original files written by the author of a project. i'd really hate accidentally committing a cache file put there by my build tool in a seemlingly harmless operation like `update`.

of course i'd be fine with a new command to put the file there, but only when i actively request sbt to so so. if i didn't, i tend to think it should go into `target/` where i can get rid of possibly wrong data with the usual `sbt clean`.

**Comment by ktham** [comment]
I don't like it being in the `target` directory either, that is for compiled artifacts and everyone gitignores that directory.

**Comment by ritschwumm** [comment]
true, the proposed "cache" file is an interesting case. neither `src/` nor `target/` really fit here - and in some case you want it checked in, and in others you don't.

**Comment by ktham** [comment]
Well, I wouldn't call it a "cache" file. I would use the PR title's language and call this a dependency "lock" file. In most cases you *do* want to commit this dependency lock file and store in source control like you would a `yarn.lock` file for JavaScript dependencies, or `Gemfile.lock` for Ruby dependencies. I'm not sure in what cases it is undesirable to check this file in.

**Comment by ritschwumm** [comment]
how would you update of your dependencies?

right now it's
* update the build.sbt
* commit

with this file in `src/` it would be
* update your build.sbt
* `sbt reload`
* `sbt update`
* commit
which is quite a bit more work.

this would be quite annoying when you - like i do for most of my projects - use a strict conflict manager and use only very few external libraries.

**Comment by ShaneDelmore** [comment]
I like the lock file in the root of the project but I can also see the argument for putting it in managed_src.

**Comment by sideeffffect** [comment]
I would suggest to do what all the other packages managers do ([`F#`](https://fsprojects.github.io/Paket/lock-file.html), [`Rust`](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html), [`Javascript`](https://yarnpkg.com/lang/en/docs/yarn-lock/), [`Ruby`](https://bundler.io/v1.3/rationale.html))

in this case of where to put the `lock` file, it would be the root of the whole project

@jvican is this something you're working on? do you think this is a feature that could ever get into `sbt`?

**Comment by jvican** [comment]
> @jvican is this something you're working on? do you think this is a feature that could ever get into sbt?
  
Not working on it, it's been a while I'm not active in sbt.

**Comment by nrdxp** [comment]
I'm interested in using SBT to build Scala programs inside of [Nix](https://nixos.org) derivations. I've left some comments on my endeavors on a plugin that already does rudimentary dependency locking for SBT https://github.com/stringbean/sbt-dependency-lock/issues/28#issuecomment-998357616

Anyway, I'd just like to say that any such lock file really should contain _everything_ necessary to build a given project, including build time SBT plugins, and even the version of SBT in use and the libraries it drops in it's boot directory on startup. 

Also, most of the lock formats I've seen include a full URL to the resolved artifact, and I agree with @sideeffffect that the lockfile should encompass all of the subprojects and just exist at the project root directory. I've got my :crossed_fingers: that something like this will make it into 2.0.

**Comment by nrdxp** [comment]
Wanted to post a quick update. I was able to successfully (albeit completely outside of sbt) compile a JSON lock file just from the information available in the Coursier cache directory that SBT builds. You can get some details and see the script here:
https://github.com/zaninime/sbt-derivation/issues/5#issuecomment-1124181516

I'm sure it wouldn't be too hard (for those with more Scala experience than myself) to spit out some very similar JSON during a run of SBT.

**Comment by ptrdom** [comment]
Is there any work being done on this? Are there any blockers?
